types/serverless-step-functions/index.d.ts

```diff
@@ -0,0 +1,310 @@
+ * This package uses module augmentation to extend the existing types
+ * from @serverless/typescript.
+ *
+ * ```
+ * import type { AWS } from '@serverless/typescript';
+ *
+ * const serverlessConfiguration: AWS = {
+ *   service: 'aws-nodejs-typescript',
+ *   frameworkVersion: '*',
+ *   provider: {
+ *     name: 'aws',
+ *     runtime: 'nodejs12.x',
+ *   },
+ *   stepFunctions: {
+ *     stateMachines: {
+ *       helloWorldStepFunction: {
+ *         name: '${self:service}-helloWorldStepFunction-${sls:stage}',
+ *         definition: {
+ *           // <your definition>
+ *         }
+ *       }
+ *     }
+ *   }
+ * }
+ * ```
+ * @see https://github.com/serverless/typescript
+ * @see https://github.com/serverless-operations/serverless-step-functions
+ */
+
+declare module '@serverless/typescript' {
+    interface AWS {
+        stepFunctions?: {
+            stateMachines: StateMachines;
+            validate?: boolean;
+            noOutput?: boolean;
+        };
+    }
+}
+
+declare module 'serverless-step-functions' {
+    interface StepFunctions {
+        stateMachines: StateMachines;
+        validate?: boolean;
+        noOutput?: boolean;
+    }
+}
+
+type StateMachines = {
+    [stateMachine: string]: StateMachine;
+};
+
+export type StateMachine = {
+    type?: 'STANDARD' | 'EXPRESS';
+    id?: string;
+    name?: string;
+    definition: Definition;
+    tracingConfig?: TracingConfig;
+    loggingConfig?: LoggingConfig;
+    events?: any[];
+    dependsOn?: AwsResourceDependsOn;
+    tags?: AwsResourceTags;
+};
+
+type LoggingConfig = {
+    level: 'ERROR' | 'ALL' | 'FATAL' | 'OFF';
+    includeExecutionData: boolean;
+    destinations: Resource | Resource[];
+};
+
+type JSONPrimitive = string | number | boolean | null;
+type JSONObject = { [member: string]: JSONValue };
+interface JSONArray extends Array<JSONValue> {}
+type JSONValue = JSONPrimitive | JSONObject | JSONArray;
+
+type Selector = string | Record<string, unknown>;
+
+type Input = {
+    'AWS_STEP_FUNCTIONS_STARTED_BY_EXECUTION_ID.$'?: string;
+} & Record<string, any>;
+
+type ResultSelector = Record<string, any>;
+
+type Parameter = {
+    'Payload.$'?: string;
+    Input?: Input;
+    FunctionName?: string;
+    StateMachineArn?: string;
+} & {
+    [selector: string]: Selector;
+};
+
+type TracingConfig = {
+    enabled: boolean;
+};
+
+type Definition = {
+    Comment?: string;
+    StartAt: string;
+    States: States;
+    ProcessorConfig?: {
+        Mode: 'INLINE' | 'DISTRIBUTED';
+    };
+};
+
+type States = {
+    [state: string]: Choice | Fail | Map | Task | Parallel | Pass | Wait | Succeed;
+};
+
+type StateBase = {
+    Catch?: Catcher[];
+    Retry?: Retrier[];
+    End?: boolean;
+    InputPath?: string;
+    Next?: string;
+    OutputPath?: string;
+    ResultPath?: string | null;
+    ResultSelector?: { [key: string]: string | { [key: string]: string } };
+    Type: string;
+    Comment?: string;
+};
```

Comment:
There are fields in this base type that are not permitted on some of the subtypes. For example, a `Choice` cannot have the `End` property.

I suggest removing this type and add the necessary fields where needed in the specific state types

[choice docs](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-choice-state.html) for reference.